### PR TITLE
ES5:Emit parameter initialiser before object rest destructuring

### LINKED ISSUE #14026: Possible bug: Complex named function params with defaults and rest causes errors.
**TypeScript Version:**  `2.1.1` & `rc`

Sorry if this has already been reported. I searched around and couldn't find exactly this.

This is sort of difficult to describe, so here's an example:

```ts
interface Foo {
    bar?: any;
    baz?: any;
}

function foobar({ bar = {}, ...opts }: Foo = {}) {
    console.log(bar);
    console.log(opts.baz);
}

foobar(); // Throws error
foobar({ baz: 'hello' }); // No error
foobar({ bar: { greeting: 'hello' } }); // No error
```

[Playground example](http://www.typescriptlang.org/play/index.html#src=interface%20Foo%20%7B%0D%0A%20%20%20%20bar%3F%3A%20any%3B%0D%0A%20%20%20%20baz%3F%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Afunction%20foobar(%7B%20bar%20%3D%20%7B%7D%2C%20...opts%20%7D%3A%20Foo%20%3D%20%7B%7D)%20%7B%0D%0A%20%20%20%20console.log(bar)%3B%0D%0A%20%20%20%20console.log(opts.baz)%3B%0D%0A%7D%0D%0A%0D%0Afoobar()%3B%20%2F%2F%20Throws%20error%0D%0Afoobar(%7B%20baz%3A%20'hello'%20%7D)%3B%20%2F%2F%20No%20error%0D%0Afoobar(%7B%20bar%3A%20%7B%20greeting%3A%20'hello'%20%7D%20%7D)%3B%20%2F%2F%20No%20error)

**Expected behavior:**
I'd expect `bar` to still be initialized and available in the function call as an empty object when no args are given to the function.

**Actual behavior:**

The following error gets thrown (reproducible in the playground example above).
```
Uncaught TypeError: Cannot read property 'bar' of undefined
    at foobar (<anonymous>:11:16)
    at <anonymous>:16:1
    at HTMLButtonElement.excuteButton.onclick (http://www.typescriptlang.org/play/playground.js:242)
```

### PR DESCRIPTION
**Pull request by sandersn**
Fix #14026, where ES5 emit for a parameter with

1. a default value initialiser
2. an object binding pattern containing an object rest

incorrectly emitted the destructuring for the object rest before the default value initialisation.

This happened because, during emit, the ES next transform runs first, transforming object rest destructuring and marking it as part of the function prologue. Then the ES5 transform runs and transforms the default initialiser, also marking it as part of the prologue. Then the prologue is emitted in the order the statements were added.

The fix is to not mark the object rest destructuring as part of the prologue. I'm not 100% sure that this is the right fix, but it fixes the bug as it stands today.

Here's an example:

```ts
function foobar({ bar={}, ...opts }: any = {}) { }
```

which should have the ES5 emit:

```js
function foobar(_a) {
  if (_a === void 0) { _a = {}; }
  var _b = _a.bar, bar = _b === void 0 ? {} : _b, opts = __rest(_a, ["bar"]);
}
```

### CODE REVIEW

**Comment by sandersn** [comment]
From in-person discussion with @rbuckton, I'm going to try the alternate fix of `unshift`ing onto the prologue instead of `push`ing.

File: src/compiler/transformers/es2015.ts [hunk 1]
```diff
@@ -915,14 +915,14 @@ namespace ts {
...
             }
-            else if (constructor) {
-                // Otherwise, try to emit all potential prologue directives first.
-                statementOffset = addPrologueDirectives(statements, constructor.body.statements, /*ensureUseStrict*/ false, visitor);
-            }
 
             if (constructor) {
                 addDefaultValueAssignmentsIfNeeded(statements, constructor);
                 addRestParameterIfNeeded(statements, constructor, hasSynthesizedSuper);
+                if (!hasSynthesizedSuper) {
```
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: We need to always emit *real* prologue directives first (e.g. "use strict"), then our custom prologue statements.

Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: Maybe we need to split `addPrologueDirectives` into two functions, `addPrologueDirectives` (which only handles *real* prologue directives) and `addCustomPrologueDirectives` (which only handles our custom leading statements).

File: src/compiler/transformers/es2015.ts [hunk 2]
```diff
@@ -1825,6 +1818,10 @@ namespace ts {
             }
 
             if (isBlock(body)) {
+                // ensureUseStrict is false because no new prologue-directive should be added.
+                // addPrologueDirectives will simply put already-existing directives at the beginning of the target statement-array
+                const statementOffset = addPrologueDirectives(statements, body.statements, /*ensureUseStrict*/ false, visitor);
```
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: This also would result in a `"use strict"` appearing after executable statements.

Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: I'm surprised we either don't have a test for this or that an existing test isn't failing.

Comment (sandersn) [inline src/compiler/transformers/es2015.ts]: I'm trying to create a test and realised I don't know how to get `"use strict"` to be generated anywhere except at the beginning of a file. How can I get it to emit at the beginning of a function?

File: src/compiler/transformers/es2015.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/transformers/es2015.ts]: How do I tell the difference? Is "use strict" the only real prologue directive right now?

File: src/compiler/transformers/es2015.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: You write it yourself. We respect existing directives.

File: src/compiler/factory.ts [hunk 3]
```diff
@@ -3058,14 +3058,26 @@ namespace ts {
...
+        return addCustomPrologue(target, source, offset, visitor);
+    }
+
+    /**
+     * Add just the standard (string-expression) prologue-directives into target statement-array.
+     * The function needs to be called during each transformation step.
+     * This function needs to be called whenever we transform the statement
+     * list of a source file, namespace, or function-like body.
+     */
+    export function addPrologueDirectives(target: Statement[], source: Statement[], ensureUseStrict?: boolean): number {
```
Comment (rbuckton) [inline src/compiler/factory.ts]: To differentiate, perhaps its name should be `addStandardPrologue`?

**Pull request merged**